### DISCUSSION TITLE
SolidMaterialProperties dependency on solid temperature in THM

### DISCUSSION THREAD

**Discussion by quentin-faure** [idaholab/moose#24156] [Q&A General] [answered]
Hey everyone,

I am using the SolidMaterialProperties (for 2D heat structures in THM) which says that the material properties can be function. I am trying to have these properties be functions of the heat structure temperatures themselves. I have a very lengthy idea of how to do it (evaluation of the temperature at each element of the heat structure using post processors and then feeding them to functions). Is there a better, quicker and more efficient way to do it?

Thanks,

**Comment by GiudGiud** [comment]
Hello

I'm not sure why post processors would be involved?
The code is already dependent on the local heat structure temperature? All the function declarations are "SolidMaterialProperties::k(const ADReal & temp) const", with an explicit temperature dependence

Guillaume

**Comment by licharlot** [comment, marked as answer]
Hi Quentin, 

As Guillaume said, you can set your properties as a function, with the variable `t`as the temperature. Then use the functions into the `SolidMaterialProperties`, see the example below. Note that there is also the option to define these properties directly in a `Materials` block, and omit the parameter `materials` in the heat structure. 


```
[Functions]
  [cp_ss_fn]
    type = PiecewiseLinear
    x = '255	283	311	533	700	811	1089	1367'
    y = '441.25	443.75	461.25	532.5	551.25	563.75	613.75	613.75'
  []
  [k_ss_fn]
    type = PiecewiseLinear
    x = '255	283	311	533	700	811	1089	1367'
    y = '12.46	13.71	14.16	17.64	20.25	22	26.36	26.36'
  []
[]

[HeatStructureMaterials]
  [steel]
    type = SolidMaterialProperties
    k = k_ss_fn
    cp = cp_ss_fn
    rho = 8000
  []
[]
```

**Comment by quentin-faure** [comment]
Hi Guillaume and Lise,

Thank you for the answers, they are very useful.

Quentin
